### DISCUSSION TITLE
HTTP remote control interface

### DISCUSSION THREAD

**Discussion by Stefan-Olt** [smplayer-dev/smplayer#456] [Q&A] [answered]
Hello,
I'm currently writing a small Python application for a specific task. For that I need a cross-plattform mediaplayer, that is remote-controlable (load file, play, pause, seek, speed adjust, volume adjust) and remote-readable (playback status, current time index). Is that somehow possible with SMPlayer? A HTTP interface would probably the easiest way, but I'm open for other ideas. 
I first tried VLC, but there is one major issue: The time is rounded to seconds, I need more exact timing (preferable frame accurate, but 0.1 seconds resolution would be enough).
Best regards
Stefan

**Comment by smplayer-dev** [comment, marked as answer]
SMPlayer provides a [mpris interface](https://specifications.freedesktop.org/mpris-spec/latest/) but it allows a very basic control.

**Comment by Stefan-Olt** [comment]
That looks great, I looked into the docs and if SMPlayer implements the features listed in the description Interface MediaPlayer2.Player that's exactly what I need. Does it only work on Linux or also on Windows and macOS?

**Comment by Stefan-Olt** [comment]
I tried using the DBus-Interface (starting with dbus-send), while I got those functions working that are defined as `public slots`, those that are just `public` I cannot call. For example this is working fine:
`dbus-send --print-reply --dest=org.mpris.MediaPlayer2.smplayer /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.PlayPause`
but this won't work:
`dbus-send --print-reply --dest=org.mpris.MediaPlayer2.smplayer /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.Rate`
What am I doing wrong? I'm sorry if that's a stupid mistake, I never worked with DBus before.

I'm pretty sure I found a bug in line 118 of mediaplayer2player.cpp, that will round the position to complete seconds:
`double secs = Position / 1000000;`
I think to not loose the precision it should be:
`double secs = (double)Position / 1000000.0;`

**Comment by smplayer-dev** [comment]
I actually only tested with qdbus and it seems to work:
```
qdbus  org.mpris.MediaPlayer2.smplayer /org/mpris/MediaPlayer2  org.mpris.MediaPlayer2.Player.Rate
1
```

**Comment by Stefan-Olt** [comment]
Thanks for answer, yes, that also works for me. But I still couldn't figure out how to do that in Python, this works fine:
```
smp = QDBusInterface(r'org.mpris.MediaPlayer2.smplayer', "/org/mpris/MediaPlayer2", "org.mpris.MediaPlayer2.Player")
smp.call("Pause")
smp.call("Seek",QDBusArgument(int(value_in_ms)*1000, QMetaType.LongLong))
```
But for getting/setting properties I cannot find an example and this doesn't work, it returns None:
`smp.property("Rate")`
I also tried calling it as a method:
`smp.call("Rate")`
But this returns as error message:
`No such method 'Rate' in interface 'org.mpris.MediaPlayer2.Player' at object path '/org/mpris/MediaPlayer2' (signature '')`
Any ideas? I'm not sure if that is Qt-specific, Python-specific or both

**Comment by smplayer-dev** [comment]
I don't know about python, but this seems to work with dbus-send:

`dbus-send --print-reply --dest=org.mpris.MediaPlayer2.smplayer /org/mpris/MediaPlayer2 org.freedesktop.DBus.Properties.Get string:'org.mpris.MediaPlayer2.Player' string:'Rate'`

**Comment by Stefan-Olt** [comment]
Ah, yes, now I see my mistake, I assumed as 'Rate' is a double have to write double:'Rate'. Still doesn't help me with Python, I'll probably have to ask at stackoverflow.
Is it possible for you to fix the bug regarding the precision mentioned before or should I create a pull request?

**Reply by smplayer-dev** [reply]
> Is it possible for you to fix the bug regarding the precision mentioned before or should I create a pull request?

I've just fixed it.

**Comment by Stefan-Olt** [comment]
I found another bug: Volume is always reported as 0.4, no matter what

**Reply by smplayer-dev** [reply]
Fixed.

**Comment by Stefan-Olt** [comment]
Does the Windows (and Mac) version contain DBus support? I compiled DBus for Windows, got it running, I exported the DBus address:
`setx DBUS_SESSION_BUS_ADDRESS tcp:host=localhost,port=54321,family=ipv4`, started SMPlayer, but dbus-send always replies with Error `org.freedesktop.DBus.Error.ServiceUnknown: The name org.mpris.MediaPlayer2.smplayer was not provided by any .service files`
dbus-send can connect to the DBus, because if I do not set the environment variable I get an error, that it cannot connect to the DBus

**Comment by smplayer-dev** [comment]
SMPlayer is compiled without qdbus support on Windows and Mac. I can try to enable it and see what happens.

**Reply by Stefan-Olt** [reply]
Thanks, that would be great!

**Comment by smplayer-dev** [comment]
Package for Mac OS built with QtBus:
https://github.com/smplayer-dev/smplayer/actions/runs/2460866203

I haven't tested it.

**Comment by smplayer-dev** [comment]
Packages for Windows:
https://github.com/smplayer-dev/smplayer/actions/runs/2461599437

**Comment by Stefan-Olt** [comment]
Thanks for the effort, but I cannot find out how to connect to the D-Bus with something else than dbus-send. Maybe QDBus needs some special options. Getting D-Bus to work on Windows and Mac doesn't seem to be easy at all.
SMPlayer contains a webserver, but that is located in another binary, so I assume it's not suited for remote control. I found a simple cross-platform http server in a single header file [(here)](https://github.com/hellerf/EmbeddableWebServer), so no additional libraries. If you're ok with it I would write a simple interface for accessing/controlling the player similar to the D-Bus interface that could be activated over the command line (something --enable-http-control 8088)

**Comment by Stefan-Olt** [comment]
I tried creating a HTTP interface, it works except for opening a new file (the file is opened, but SMPlayer becomes unresponsive). I assume that's because the function is called from another thread. Any advice how to open a new file from another thread?
For QDBus, I tried to figure out how `QDBusConnection::connectToBus(const QString &address, const QString &name)` works and what  `address` and `name` should look like, but the documentation is rather limited, the source is just passing the parameters to the next function and the only source I found for that is generated from something else and not helpful at all.

Edit: I created a second custom D-Bus on Linux, and I can connect to it just fine passing the address `tcp:host=localhost,port=54321,family=ipv4` (I verified it actually connects to my custom DBus). The same doesn't work on Windows or Mac, even though I use the very same config file to start dbus-daemon.
